chore: improve topos-tce-storage documentation (#353)
topos-protocol/topos: crates/topos-tce-storage/README.md

# topos-tce-storage

The library provides the storage layer for the Topos TCE.

> It is responsible for storing and retrieving the certificates, managing the

Author: I don't have a strong opinion on that, I was afraid that having too many things as links or capitalized could lead to less readable doc. I don't know, honestly.

> It is responsible for storing and retrieving the certificates, managing the

Reviewer: Just linking to the official docs I think is enough.

So we can lead with bullet points after or before, or just link the words which we didn't explain.

Certificates: https://docs.topos.technology/content/module-1/4-protocol.html

TCE: https://docs.topos.technology/content/module-1/4-protocol.html

pending certificates pool and the certificate status, storing the different
metadata related to the protocol and the internal state of the TCE.

The storage layer is implemented using RocksDB.
The library is exposing multiple store that are used by the TCE.

### Architecture

The storage layer is composed of multiple stores that are used by the TCE.
Each store is describe in detail in its own module.

As an overview, the storage layer is composed of the following stores:

> <picture>

Reviewer 2: Picture is a bit confusing `ValidatorPerEpochStore` and `EpochValidatorStore`? 
`ValidatorPerEpochStore` is not mentioned in the text

Author: It is not use for now as we only have on active `epoch` yes

 <source media="(prefers-color-scheme: dark)" srcset="https://github.com/topos-protocol/topos/assets/1394604/5bb3c9b1-ac5a-4f59-bd14-29a02163272e">
 <img alt="Text changing depending on mode. Light: 'So light!' Dark: 'So dark!'" src="https://github.com/topos-protocol/topos/assets/1394604/e4bd859e-2a6d-40dc-8e84-2a708aa8a2d8">
</picture>

#### Definitions and Responsibilities

As illustrated above, multiple `stores` are exposed in the library using various `tables`.

The difference between a `store` and a `table` is that the `table` is responsible for storing

> the data while the `store` is responsible for managing the data and its access and behaviour.

Author: Dealing with the `pending` when delivering a certificate or managing the locks when inserting in multiple tables.

Here's the list of the different stores and their responsibilities:

- The [`EpochValidatorsStore`](struct@epoch::EpochValidatorsStore) is responsible for managing the list of validators for each `epoch`.
- The [`FullNodeStore`](struct@fullnode::FullNodeStore) is responsible for managing all the persistent data such as [`Certificate`] delivered and associated `streams`.
- The [`IndexStore`](struct@index::IndexStore) is responsible for managing indexes in order to collect information about the broadcast and the network.
- The [`ValidatorStore`](struct@validator::ValidatorStore) is responsible for managing the pending certificates pool and all the transient and volatile data.

For more information about a `store`, see the related doc.

Next, we've the list of the different tables and their responsibilities:

- The [`EpochValidatorsTables`](struct@epoch::EpochValidatorsTables) is responsible for storing the list of validators for each `epoch`.

> - The [`ValidatorPerpetualTables`](struct@validator::ValidatorPerpetualTables) is responsible for storing the [`Certificate`] delivered and all the persitent data related to the broadcast.

Author: It is the opposit of the `per epoch` (which isn't implemented because we only have on epoch for now)

- The [`ValidatorPendingTables`](struct@validator::ValidatorPendingTables) is responsible for storing the pending certificates pool and all the transient and volatile data.

> - The [`IndexTables`](struct@index::IndexTables) is responsible for storing indexes about the delivery of [`Certificate`] such as `target subnet stream`.

Author: No I mean that the `IndexTables` stores indexes, like the `Target subnet` stream.

### Special Considerations

When using the storage layer, you need to be aware of the following:
- The storage layer is using [rocksdb](https://rocksdb.org/) as a backend, which means that this storage doesn't need external service, as `rocksdb` is embeddable kv store.
- The storage layer is using [`Arc`](struct@std::sync::Arc) to share the stores between threads. It also means that a `store` is only instantiated once.

> - Some functions are batching multiple writes in one transaction. But not all functions are using it.

Author: I think this is mostly covered by `RocksDB` itself in its documentation.

### Design Philosophy

The choice of using [rocksdb](https://rocksdb.org/) as a backend was made because it is a well known and battle tested database.
It is also very fast and efficient when it comes to write and read data.

Multiple `stores` and `tables` exists in order to allow admin to deal with backups or
snapshots as they see fit. You can pick and choose which `tables` you want to backup without having to backup the whole database.

By splitting the data in dedicated tables we define strong separation of concern
directly in our storage.

`RocksDB` is however not the best fit when it comes to compose or filter data based on the data
itself.

For complex queries, another database like [`PostgreSQL`](https://www.postgresql.org/) or [`CockroachDB`](https://www.cockroachlabs.com/) could be used as a Storage for projections.

> The source of truth would still be [rocksdb](https://rocksdb.org/) but the projections would be stored in a relational database, allowing for more complex queries.

Author: I think it does, a projection database is not meant to be the SSOT of any data. It can be deleted and rebuilt from scratch without issues, and losing the projection database should lead to any data loss.

The point here is to explain why some capabilities, like having complex queries, are not meant to be part of the different stores that rely on `RocksDB`. 
It also points that we could provide a dedicated relational database layer implementation in order to provide those queries for the different API.

It is mainly here to prevent people asking or searching how to do things that is not possible for those implementations.

> The source of truth would still be [rocksdb](https://rocksdb.org/) but the projections would be stored in a relational database, allowing for more complex queries.

Reviewer: I have to say I agree with @dvdplm here. This is not a paper which discusses possibilities. The target audience is people who want to contribute to the project, and want a quick read on what piece of code means what. 

Mentioning other solutions can lead to being outdated faster, discussions around it etc. etc. 

To have this information is valid, I just don't think docs in a repository are the place to have it. One example: They might not have this question but another idea. Like how many should we guess and what should we cover? 

We should just cover what we have. Every discussions or anything else, people can open an issue on GitHub and we can discuss it there.

> The source of truth would still be [rocksdb](https://rocksdb.org/) but the projections would be stored in a relational database, allowing for more complex queries.

Reviewer: Comments like this is exactly why we shouldn't argue for or against a solution in documentation. 
https://github.com/topos-protocol/topos/pull/353/files#r1383286732

Because it never ends :) There is always something not accurately portrait etc.

> The source of truth would still be [rocksdb](https://rocksdb.org/) but the projections would be stored in a relational database, allowing for more complex queries.

Reviewer 3: Ideally/Eventually we could refer to the RFC that introduced this decision which I think explained the rationales and the alternatives, although the current RFC is bit old and outdated on some parts (and also the repo is still private)

In the meantime, I'm good with either removing this part, or squashing by having not more than an one-liner e.g., "Our state representation along with the ways we are accessing it makes RocksDB pretty suitable in our context."

As mention above, the different stores are using [`Arc`](struct@std::sync::Arc), allowing a single store to be instantiated once

> and then shared between threads. This is very useful when it comes to the [`FullNodeStore`](struct@fullnode::FullNodeStore) as it is used in various places but need to provides single entrypoint to the data.

Reviewer 2: ```suggestion
and then shared between threads. This is very useful when it comes to the [`FullNodeStore`](struct@fullnode::FullNodeStore) as it is used in various places, but should provide a single entry point for data.
```

> It also means that the store is immutable, which is a good thing when it comes to concurrency.

Author: I updated this part

The burden of managing the locks is handled by the [`async_trait`](https://docs.rs/async-trait/0.1.51/async_trait/) crate when using the [`WriteStore`](trait@store::WriteStore).

> The locks are responsible for preventing any other query to mutate the data currently in processing. For more information about the locks see [`locking`](module@fullnode::locking)

Author: yes the links are made for rust doc generate html, the README is just a quick win to display some information

- The [`IndexTables`](struct@index::IndexTables) is responsible for storing indexes about the delivery of [`Certificate`]s such as `target subnet stream`.

### Special Considerations

When using the storage layer, be aware of the following:
- The storage layer uses [rocksdb](https://rocksdb.org/) as the backend, which means don't need an external service, as `rocksdb` is an embedded key-value store.
- The storage layer uses [`Arc`](struct@std::sync::Arc) to share the stores between threads. It also means that a `store` is only instantiated once.
- Some storage methods are batching multiple writes into a single transaction.

### Design Philosophy

The choice of using [rocksdb](https://rocksdb.org/) as a backend was made because it matches a lot of the conditions
that we were expected, such as being embedded and having good performances when reading and
writing our data.

Splitting storage into multiple `stores` and `tables` allows admins to deal with backups or
snapshots as they see fit. You can pick and choose which `tables` you want to backup without having to backup the whole database.

> Strong separation of converns is directly achieved at the storage level by splitting the data in

Reviewer 3: typo still there apparently
